### PR TITLE
Add setup check to verify that the used DB version is still supported…

### PR DESCRIPTION
**Pull request by MorrisJobke**
… in the next major release

### CODE REVIEW

**Comment by MorrisJobke** [comment]
See https://github.com/nextcloud/documentation/pull/5895

File: apps/settings/lib/SetupChecks/SupportedDatabase.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	public function __construct(IL10N $l10n, IDBConnection $connection) {
+		$this->l10n = $l10n;
+		$this->connection = $connection;
+
+		$this->check();
+	}
+
+	public function check() {
+		switch (get_class($this->connection->getDatabasePlatform())) {
+			case MySQL80Platform::class: # extends MySQL57Platform
```
Comment (MorrisJobke) [inline apps/settings/lib/SetupChecks/SupportedDatabase.php:57]: I left all the platforms in here to have the full overview (except DB2 or MSSQLServer as we don't support them at all)

File: apps/settings/lib/SetupChecks/SupportedDatabase.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+				$version = strtolower($row['Value']);
+
+				if (strpos($version, 'mariadb') !== false) {
+					if (version_compare($version, '10.4', '<')) {
+						$this->description = $this->l10n->t('MariaDB version "%s" is used. Nextcloud 21 will no longer support this version and requires MariaDB 10.4 or higher.', $row['Value']);
+						return;
+					}
+				} else {
+					if (version_compare($version, '8', '<')) {
+						$this->description = $this->l10n->t('MySQL version "%s" is used. Nextcloud 21 will no longer support this version and requires MariaDB 8 or higher.', $row['Value']);
```
Comment (nickvergessen) [inline apps/settings/lib/SetupChecks/SupportedDatabase.php:73]:
```suggestion
						$this->description = $this->l10n->t('MySQL version "%s" is used. Nextcloud 21 will no longer support this version and requires MySQL 8 or higher.', $row['Value']);
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: apps/settings/lib/SetupChecks/SupportedDatabase.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	/** @var IDBConnection */
+	private $connection;
+
+	private $description = '';
+
+	public function __construct(IL10N $l10n, IDBConnection $connection) {
+		$this->l10n = $l10n;
+		$this->connection = $connection;
+
+		$this->check();
```
Comment (ChristophWurst) [inline apps/settings/lib/SetupChecks/SupportedDatabase.php:52]: I'm not a fan of running actual class logic from the constructor. can we do this in `run`?

Comment (MorrisJobke) [reply apps/settings/lib/SetupChecks/SupportedDatabase.php:52]: Then it depends on the fact that `run()` is called first, before fetching description for example.

Comment (ChristophWurst) [reply apps/settings/lib/SetupChecks/SupportedDatabase.php:52]: Yes

Comment (MorrisJobke) [reply apps/settings/lib/SetupChecks/SupportedDatabase.php]: Fixed

**Comment by MorrisJobke** [comment]
Ready for review :)

File: apps/settings/lib/SetupChecks/SupportedDatabase.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			case MySQL57Platform::class: # extends MySQLPlatform
+			case MariaDb1027Platform::class: # extends MySQLPlatform
+			case MySqlPlatform::class:
+				$result = $this->connection->prepare('SHOW VARIABLES LIKE "version";');
+				$result->execute();
+				$row = $result->fetch();
+				$version = strtolower($row['Value']);
+
+				if (strpos($version, 'mariadb') !== false) {
+					if (version_compare($version, '10.4', '<')) {
```
Comment (rullzer) [inline apps/settings/lib/SetupChecks/SupportedDatabase.php:71]:
```suggestion
					if (version_compare($version, '10.2', '<')) {
```

**Review by rullzer — CHANGES_REQUESTED** [review]

File: apps/settings/lib/SetupChecks/SupportedDatabase.php (hunk 4 above)
Comment (nickvergessen) [reply apps/settings/lib/SetupChecks/SupportedDatabase.php:71]: But 10.4 is the mysql8 compatible version?
Ref https://github.com/nextcloud/server/blob/stable16/lib/private/Setup/MySQL.php#L48-L56

Comment (MorrisJobke) [reply apps/settings/lib/SetupChecks/SupportedDatabase.php:71]: Our docs say, that 10.2 works and CI runs against 10.1 🙈

Comment (MorrisJobke) [reply apps/settings/lib/SetupChecks/SupportedDatabase.php]: I changed it to 10.2 for now. We can raise it later again, okay?

**Comment by MorrisJobke** [comment]
@rullzer @ChristophWurst Ready for review :)

File: apps/settings/lib/SetupChecks/SupportedDatabase.php (hunk 4 above)
Comment (rullzer) [reply apps/settings/lib/SetupChecks/SupportedDatabase.php]: 10.2 is still supported so yeah :)

**Review by rullzer — APPROVED** [review]
Looks good.
But didn't test

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**